src/Microsoft.ML.Data/Transforms/ValueToKeyMappingTransformer.cs

```diff
@@ -789,38 +789,86 @@ private bool SaveAsOnnxCore(OnnxContext ctx, int iinfo, ColInfo info, string src
                     var terms = GetTermsAndIds<ReadOnlyMemory<char>>(iinfo, out termIds);
                     node.AddAttribute("keys_strings", terms);
                 }
-                else if (info.TypeSrc.GetItemType().Equals(NumberDataViewType.Single))
+                else if (type.Equals(NumberDataViewType.Single))
                 {
                     node = ctx.CreateNode(opType, srcVariableName, labelEncoderOutput, ctx.GetNodeName(opType));
                     var terms = GetTermsAndIds<float>(iinfo, out termIds);
                     node.AddAttribute("keys_floats", terms);
                 }
-                else if (info.TypeSrc.GetItemType().Equals(NumberDataViewType.Double))
+                else if (type.Equals(NumberDataViewType.Double))
                 {
                     // LabelEncoder doesn't support double tensors, so values are cast to floats
                     var castOutput = ctx.AddIntermediateVariable(NumberDataViewType.Single, "castOutput", true);
-                    castNode = ctx.CreateNode("Cast", srcVariableName, castOutput, ctx.GetNodeName(opType), "");
+                    castNode = ctx.CreateNode("Cast", srcVariableName, castOutput, ctx.GetNodeName("Cast"), "");
                     var t = InternalDataKindExtensions.ToInternalDataKind(DataKind.Single).ToType();
                     castNode.AddAttribute("to", t);
                     node = ctx.CreateNode(opType, castOutput, labelEncoderOutput, ctx.GetNodeName(opType));
                     var terms = GetTermsAndIds<double>(iinfo, out termIds);
                     node.AddAttribute("keys_floats", terms);
                 }
-                else if (info.TypeSrc.GetItemType().Equals(NumberDataViewType.Int64))
+                else if (type.Equals(NumberDataViewType.Int64))
                 {
                     // LabelEncoder doesn't support mapping int64 -> int64, so values are cast to strings
                     var castOutput = ctx.AddIntermediateVariable(TextDataViewType.Instance, "castOutput", true);
-                    castNode = ctx.CreateNode("Cast", srcVariableName, castOutput, ctx.GetNodeName(opType), "");
+                    castNode = ctx.CreateNode("Cast", srcVariableName, castOutput, ctx.GetNodeName("Cast"), "");
                     var t = InternalDataKindExtensions.ToInternalDataKind(DataKind.String).ToType();
                     castNode.AddAttribute("to", t);
                     node = ctx.CreateNode(opType, castOutput, labelEncoderOutput, ctx.GetNodeName(opType));
-                    var terms = GetTermsAndIds<long>(iinfo, out termIds);
+                    var terms = GetTermsAndIds<Int64>(iinfo, out termIds);
+                    node.AddAttribute("keys_strings", terms.Select(item => item.ToString()));
+                }
+                else if (type.Equals(NumberDataViewType.Int32))
+                {
+                    var castOutput = ctx.AddIntermediateVariable(TextDataViewType.Instance, "castOutput", true);
```

Comment:
I am approving the PR but I have a concern. 
Setting the last parameter to true skips adding type and shape information. That has been causing a lot of problems and also makes it harder to debug. If this PR fixes AveragedPerceptron then please go ahead and merge this. But please go back and use the default value of false to all AddIntermediateVariable calls here and figure out what errors we see and fix those.

Reply:
Done.